File: src/main/java/com/github/weewar/mapviewer/service/impl/ImageRepositoryImpl.java
package com.github.weewar.mapviewer.service.impl;

import com.github.weewar.mapviewer.exceptions.ImageNotFoundException;
import com.github.weewar.mapviewer.exceptions.ImagePreloadException;
import com.github.weewar.mapviewer.model.enums.Direction;
import com.github.weewar.mapviewer.model.enums.Owner;
import com.github.weewar.mapviewer.model.enums.TerrainType;
import com.github.weewar.mapviewer.model.enums.UnitType;
import com.github.weewar.mapviewer.service.ImageRepository;
import com.github.weewar.mapviewer.service.impl.keys.TerrainImageKey;
import com.github.weewar.mapviewer.service.impl.keys.UnitImageKey;
import org.springframework.stereotype.Service;
import org.yaml.snakeyaml.Yaml;

import javax.imageio.ImageIO;
import java.awt.*;
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@Service
public class ImageRepositoryImpl implements ImageRepository {
    private final Map<TerrainImageKey, Image> terrainImages = new ConcurrentHashMap<>(50 /*max terrain images*/,
            0.75f /*default load factor*/, 1 /*single shard*/);
    private final Map<UnitImageKey, Image> unitImages = new ConcurrentHashMap<>(200 /*max unit images*/,
            0.75f /*default load factor*/, 1 /*single shard*/);

    @Override
    public Image getTerrain(TerrainType terrainType, Owner owner, Direction direction) throws ImageNotFoundException {
        Image image = terrainImages.get(new TerrainImageKey(terrainType, owner, direction));
        if (image == null) {
            throw new ImageNotFoundException("Terrain image not found: type = " + terrainType + ", owner = " + owner +
                    ", direction = " + direction);
        }
        return image;
    }

    @Override
    public Image getUnit(UnitType unitType, Owner owner) throws ImageNotFoundException {
        Image image = unitImages.get(new UnitImageKey(unitType, owner));
        if (image == null) {
            throw new ImageNotFoundException("Unit image not found: type = " + unitType + ", owner = " + owner);
        }
        return image;
    }

    @Override
    public void preloadImages() throws ImagePreloadException {
        try {
            Yaml yaml = new Yaml();
            preloadTerrainImages(yaml);
            preloadUnitImages(yaml);
        } catch (IOException e) {
            throw new ImagePreloadException(e);
        }
    }

    @SuppressWarnings("unchecked")
    private void preloadTerrainImages(Yaml yaml) throws IOException {
        try (InputStream inputStream = getClass().getResourceAsStream("/.terrain.yml")) {
            Map<String, Map<String, String>> terrainFiles = yaml.loadAs(inputStream, Map.class);
            for (String terrainFile : terrainFiles.keySet()) {
                Map<String, String> terrainDetails = terrainFiles.get(terrainFile);
                TerrainType terrainType = TerrainType.of(terrainDetails.get("type"));
                Owner owner = Owner.of(terrainDetails.get("owner"));
                Direction direction = Direction.of(terrainDetails.get("direction"));
                URL url = getClass().getResource(terrainFile);
                Image terrainImage = ImageIO.read(url);
                terrainImages.put(new TerrainImageKey(terrainType, owner, direction), terrainImage);
            }
        }
    }

    @SuppressWarnings("unchecked")
    private void preloadUnitImages(Yaml yaml) throws IOException {
        try (InputStream inputStream = getClass().getResourceAsStream("/.units.yml")) {
            Map<String, Map<String, String>> unitFiles = yaml.loadAs(inputStream, Map.class);
            for (String unitFile : unitFiles.keySet()) {
                Map<String, String> unitDetails = unitFiles.get(unitFile);
                UnitType unitType = UnitType.of(unitDetails.get("type"));
                Owner owner = Owner.of(unitDetails.get("owner"));
                URL url = getClass().getResource(unitFile);
                Image unitImage = ImageIO.read(url);
                unitImages.put(new UnitImageKey(unitType, owner), unitImage);
            }
        }
    }
}

File: src/main/kotlin/com/github/weewar/mapviewer/service/impl/ImageRepositoryImpl.kt
package com.github.weewar.mapviewer.service.impl

import com.github.weewar.mapviewer.model.*
import com.github.weewar.mapviewer.model.enums.*
import com.github.weewar.mapviewer.service.ImageRepository
import com.github.weewar.mapviewer.service.impl.keys.TerrainImageKey
import com.github.weewar.mapviewer.service.impl.keys.UnitImageKey
import org.springframework.stereotype.Service
import org.yaml.snakeyaml.Yaml
import java.awt.Image
import java.awt.image.BufferedImage
import java.io.IOException
import java.io.InputStream
import java.util.concurrent.ConcurrentHashMap
import javax.imageio.ImageIO

@Service
class ImageRepositoryImpl : ImageRepository {

    private val terrainImages = ConcurrentHashMap<TerrainImageKey, Image>(50, 0.75f, 1)
    private val unitImages = ConcurrentHashMap<UnitImageKey, Image>(50, 0.75f, 1)

    @Throws(ImageNotFoundException::class)
    override fun getTerrain(terrainType: TerrainType, owner: Owner?, direction: Direction?): Image =
            terrainImages[TerrainImageKey(terrainType, owner, direction)]
                    ?: throw ImageNotFoundException("Terrain image not found: type = $terrainType, owner = $owner, direction = $direction")

    @Throws(ImageNotFoundException::class)
    override fun getUnit(unitType: UnitType, owner: Owner): Image =
            unitImages[UnitImageKey(unitType, owner)]
                    ?: throw ImageNotFoundException("Unit image not found: type = $unitType, owner = $owner")

    @Throws(ImagePreloadException::class)
    override fun preloadImages() =
            try {
                preloadTerrainImages()
                preloadUnitImages()
            } catch (e: IOException) {
                throw ImagePreloadException(e)
            }

    @Throws(IOException::class)
    private fun preloadTerrainImages() = withConfigFile(terrainImagesConfig) { terrainImage, terrainInfo ->
        val terrainType = TerrainType.of(terrainInfo["type"]!!)
        val owner = Owner.of(terrainInfo["owner"])
        val direction = Direction.of(terrainInfo["direction"])
        terrainImages[TerrainImageKey(terrainType, owner, direction)] = terrainImage
    }

    @Throws(IOException::class)
    private fun preloadUnitImages() = withConfigFile(unitImagesConfig) { unitImage, unitInfo ->
        val unitType = UnitType.of(unitInfo["type"])
        val owner = Owner.of(unitInfo["owner"])
        unitImages[UnitImageKey(unitType!!, owner!!)] = unitImage
    }

    @Throws(IOException::class)
    private fun withConfigFile(configFile: String, configEntryConsumer: (image: BufferedImage, imageInfo: Map<String, String?>) -> Unit) {
        javaClass.getResourceAsStream(configFile).use { inputStream ->
            val config = loadConfig(inputStream)
            for (file in config.keys) {
                val imageInfo = config[file] as Map<String, String?>
                val image = ImageIO.read(javaClass.getResource(file))
                configEntryConsumer(image, imageInfo)
            }
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun loadConfig(inputStream: InputStream): Map<String, Map<String, String?>> =
            Yaml().loadAs(inputStream, Map::class.java) as Map<String, Map<String, String?>>
}
